controllers/volumereplicationgroup_controller.go

```diff
@@ -409,94 +420,229 @@ func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
 	return ctrl.Result{}, nil
 }
 
-func (v *VRGInstance) addFinalizer(finalizer string) error {
-	if !containsString(v.instance.ObjectMeta.Finalizers, finalizer) {
-		v.instance.ObjectMeta.Finalizers = append(v.instance.ObjectMeta.Finalizers, finalizer)
-		if err := v.reconciler.Update(v.ctx, v.instance); err != nil {
-			v.log.Error(err, "Failed to add finalizer", "finalizer", finalizer)
+// reconcileVRsAsPrimary creates/updates VolumeReplication CR for each pvc
+// from pvcList. If it fails (even for one pvc), then requeue is set to true.
+func (v *VRGInstance) reconcileVRsAsPrimary() bool {
+	requeue := false
 
-			return fmt.Errorf("failed to add finalizer to VolumeReplicationGroup resource (%s/%s), %w",
-				v.instance.Name, v.instance.Namespace, err)
+	for idx := range v.pvcList.Items {
+		pvc := &v.pvcList.Items[idx]
+		pvcNamespacedName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
+		log := v.log.WithValues("pvc", pvcNamespacedName.String())
+
+		if requeueResult := v.preparePVCForVRProtection(pvc, log); requeueResult {
+			requeue = true
+
+			continue
+		}
+
+		// Find VR for PVC
+		if err := v.processVRAsPrimary(pvcNamespacedName, log); err != nil {
+			log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
+				"errorValue", err)
+
+			requeue = true
+
+			continue
+		}
+
+		if requeueResult := v.processPVCAsVRProtected(pvc, log); requeueResult {
+			requeue = true
 		}
 	}
 
-	return nil
+	// TODO: updateVRStatus even when resource requires a requeue
+
+	return requeue
 }
 
-// handlePersistentVolumeClaims creates VolumeReplication CR for each pvc
-// from pvcList. If it fails (even for one pvc), then requeue is set to true.
-// For now, keeping creation of VolumeReplication CR and backing up of PV
-// metadata in separate functions and calling them separately. In future,
-// if creation of VolumeReplication CR and backing up of pv metadata should
-// happen together (i.e for each pvc create VolumeReplication CR and then
-// backup corresponding PV metadata), then it can be put in a single function.
-func (v *VRGInstance) handlePersistentVolumeClaims() bool {
-	requeue := true
-
-	if requeueResult := v.createVolumeReplicationResources(); requeueResult {
-		v.log.Info("Requeuing resource")
+// preparePVCForVRProtection processes prerequisites of any PVC that needs VR protection
+func (v *VRGInstance) preparePVCForVRProtection(pvc *corev1.PersistentVolumeClaim,
+	log logr.Logger) bool {
+	const requeue bool = true
+
+	// if the PVC is not bound yet, don't proceed
+	if pvc.Status.Phase != corev1.ClaimBound {
+		log.Info("Requeuing as PersistentVolumeClaim is not bound", "pvcPhase", pvc.Status.Phase)
 
 		return requeue
```

Comment:
Good point.

Currently in the compare code we are not looking at Status changes, and hence will miss it. BUT, I have a TODO in the pr #43  to add compare of finalizers (which is also absent at preset) and this [comment](https://github.com/RamenDR/ramen/pull/43#discussion_r614144011). So we can optimize this out at that point, as a status change of this field would cause a reconcile retrigger in the future.